lib/puppet/util/windows/process.rb

```diff
@@ -97,6 +121,32 @@ def with_process_token(access, &block)
   end
   module_function :with_process_token
 
+  def get_process_image_name_by_pid(pid)
+    image_name = ""
+
+     open_process(PROCESS_QUERY_INFORMATION, false, pid) do |phandle|
+
+       FFI::MemoryPointer.new(:dword, 1) do |exe_name_length_ptr|
+        # Add 1 for the null terminator, and UTF is 2 bytes/char:
+        max_path_length = (MAX_PATH_LENGTH + 1) * 2
+        exe_name_length_ptr.write_dword(max_path_length)
+        FFI::MemoryPointer.new(max_path_length) do |exe_name_ptr|
+          use_win32_path_format = 0
+          result = QueryFullProcessImageNameW(phandle, use_win32_path_format, exe_name_ptr, exe_name_length_ptr)
+          if result == FFI::WIN32_FALSE
+            raise Puppet::Util::Windows::Error.new(
+              "QueryFullProcessImageNameW(phandle, #{use_win32_path_format}, " +
+              "exe_name_ptr, #{max_path_length}")
+          end
+          image_name = exe_name_ptr.read_wide_string(MAX_PATH_LENGTH + 1)
```

Comment:
`exe_name_length_ptr` is an in/out parameter, so `QueryFullProcessImageNameW`writes the path to `exe_name_ptr` and the number of characters written to `exe_name_length_ptr`. You'll want to pass the latter to `read_wide_string`

```ruby
 exe_name_ptr.read_wide_string(exe_name_length_ptr.read_dword)
```